docs/proposals/20220125-ipam-integration.md

```diff
@@ -0,0 +1,410 @@
+
+Therefore, IPAM providers are required to bring a custom IP Pool resource. The resource can have an arbitrary structure with the specific parameters for the IPAM system in use, and by inspecting the API group and kind of the pool reference held by IPAddressClaims, the provider can determine whether it is supposed to handle the claim or not.
+
+When an IPAM provider finds a IPAddressClaim that references one of its Pools, it allocates an IP address, creates an IPAddress resource and updates the status of the IPAddressClaim with a reference to this IPAddress.
+
+A simple example for an IP Pool could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InClusterIPPool
+metadata:
+  name: some-pool
+spec:
+  pools:
+    - subnet: 10.10.10.0/24
+      start: 10.10.10.100
+      end: 10.10.10.200
+```
+A pool for an Infoblox Provider could look like this:
+
+```yaml
+apiVersion: ipam.cluster.x-k8s.io/v1alpha1
+kind: InfobloxIPPool
+metadata:
+  name: ib-pool
+spec:
+  networkView: "some-view"
+  dnsZone: "test.example.com."
+  network: 10.10.10.0/24
+```
+
+### Consumption
+
+The consumers of this API contract are the infrastructure providers. As mentioned in the introduction, not all providers will require integration with IPAM systems. In addition, DHCP may be a viable or even better option in some environments. Integration with this API should therefore be implemented as an optional feature.
+
+In order to consume IPAddresses, infrastructure providers need to create an IPAddressClaim for each address they require, and then have to wait until all claims are fulfilled. When creating those claims, a deterministic naming scheme should be used, so it's easy to identify which claims belong to which machine, e.g. `<infra machine name>-<interface name>`.
+
+Since IPAddressClaims have to reference a Pool, the infrastructure provider needs to allow specifying a Pool in its infrastructure templates, and the referenced Pool also has to be created (or exist already) when creating a cluster. The pool should be in the same namespace.
+
+![relationships of ipam custom resources](images/ipam-integration/consumption.png)
+
+When creating an IPAddressClaim, the provider should set an owner reference to the infrastructure machine, so IPAM providers can fetch further information if necessary.
+
+Using CAPV as an example, a template *could* look like this:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha3
+kind: VSphereMachineTemplate
+metadata:
+ name: example
+ namespace: vsphere-site1
+spec:
+ template:
+   spec:
+     cloneMode: FullClone
+     numCPUs: 8
+     memoryMiB: 8192
+     diskGiB: 45
+     network:
+       devices:
+       - dhcp4: false
+         fromPool: # reference to the pool
+           group: ipam.cluster.x-k8s.io/v1alpha1
+           kind: IPPool
+           name: testpool
```

Comment:
I know there was already some back and forth, but given the fact this is getting relevant from a growing number of providers, I'm wondering if we should move this to Machine level (we can also get there incrementally, but why not now?)
@vincepri @CecileRobertMichon @enxebre @sbueringer @yastij opinions?

Reply:
Not opposed though based on the journey that took us from v1alpha1 to v1alpha2 where we figured the new abstractions as CRDs and the Refs and we are still evolving conditions, I think It'd be good to see some real usage to gain some knowledge, see a few ipam implementations and how's adoption from the providers before we figure the best way to plumb it into core API.

Reply:
That would certainly make this proposal easier. Claims could be created by CAPI. Providers would still be required to fetch IPAddresses though, if they are used.
The hard part probably is figuring out all available parameters that can be set on Machine level, and how to add additional parameters on provider side in a good way.

Reply:
@vincepri @CecileRobertMichon @sbueringer @yastij opinions?

Reply:
This would be definitely save the providers from doing one-off code changes for creating the Claim objects. I will group this as an enhancement in the similar bucket as failure domains. Specifying the IPPool will follow the Ref conventions seen all through CAPI objects.
Plus it makes more sense for Machine objects to create the `IPClaim`s as the proposal intends to introduce these API types as part of the CAPI codebase.

Reply:
Hm given that we need the CRDs and at least a webhook in core CAPI already I lean towards making it a part of core CAPI so we can kind of make sure that everything works as intended (more or less like we already handle other contracts like the InfrastructureMachine where we have a machine controller which coordinates)